### PR TITLE
Implemented delivery report

### PR DESCRIPTION
**Pull request by Cogno-IDU**
Implemented delivery report system using method overloading.
When a message is sent you can decide (thanks to method overloading) to:
- only send the message
- send the message with a sent report
- send the message with a delivered report
- send the message with both the sent and delivered report

Sent and Delivered reports are created automatically when the user passes listeners for sent and delivered report.

### CODE REVIEW

File: smslibrary/src/main/java/com/eis/smslibrary/SMSMessage.java [hunk 1]
```diff
@@ -87,7 +87,7 @@ public SMSPeer getPeer() {
      * so that one can see all the possible values without having to
      * look at the official Android documentation
      * <p>
-     * These are used in {@link SMSMessageSentListener}
+     * These are used in {@link SMSSentBroadcastReceiver}
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/smslibrary/SMSMessage.java]: These states are used in the listener, the user doesn't have to care about the BroadcastReceiver, he just has to know that these states are passed as parameter inside the listener and he can use them there

File: smslibrary/src/main/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiver.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * so every IntentFilter name has to be different
+ *
+ * @author Marco Cognolato
+ */
+public class SMSDeliveredBroadcastReceiver extends BroadcastReceiver {
+    private SMSDeliveredListener listener;
+    private SMSMessage message;
+
+    /**
+     * Constructor for the custom {@link BroadcastReceiver}.
```
[the remark this answers is not in the document]
Comment (Cogno-IDU) [reply smslibrary/src/main/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiver.java]: mmm, I wonder... 😂

File: smslibrary/src/main/java/com/eis/smslibrary/SMSHandler.java [hunk 3]
```diff
@@ -98,10 +100,38 @@ public void sendMessage(final @NonNull SMSMessage message) {
      */
     public void sendMessage(final @NonNull SMSMessage message, final @Nullable SMSSentListener sentListener) {
         checkSetup();
-        String smsContent = SMSMessageHandler.getInstance().parseData(message);
         PendingIntent sentPI = setupNewSentReceiver(message, sentListener);
+        SMSCore.sendMessage(getSMSContent(message), message.getPeer().getAddress(), sentPI, null);
```
[the remark this answers is not in the document]
Comment (Cogno-IDU) [reply smslibrary/src/main/java/com/eis/smslibrary/SMSHandler.java]: Uuuh, you're right, didn't think about that! Fixing now 👍

File: smslibrary/src/main/java/com/eis/smslibrary/SMSMessage.java (hunk 1 above)
Comment (Cogno-IDU) [reply smslibrary/src/main/java/com/eis/smslibrary/SMSMessage.java]: Alright, since it wasn't something I worked on and since the link had an error (due to a class rename), I just put the class where the states were used. I'll update it now 👍

**Comment by Cogno-IDU** [comment]
I'll wait for a third review, since we have a working release of the librery this can take longer without any problem

File: smslibrary/src/main/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiver.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+     *
+     * @param message  that will be sent.
+     * @param listener to be called when the operation is completed.
+     */
+    SMSDeliveredBroadcastReceiver(@NonNull final SMSMessage message, @NonNull final SMSDeliveredListener listener) {
```
Comment (EnricoCestaro) [inline smslibrary/src/main/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiver.java]: @NonNull means that it cannot be Null (duh), but why is it final?

Comment (Cogno-IDU) [reply smslibrary/src/main/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiver.java]: The `@NonNull final` notation comes from the SMSSentBroadcastReceiver methods. Basically having an input parameter final means that whiting that method, you can't assign it new values.

Consider this example (taken from [here](https://stackoverflow.com/questions/5607773/change-a-functions-arguments-values) )
```
public class Test {

    public static void main(String[] args) {
        boolean in = false;
        truifier(in);
        System.out.println("in is " + in);
    }

    public static void truifier (boolean bool) {
        if (bool == false) {
            bool = true;
        }
        System.out.println("bool is " + bool);
    }
}
```

output:
```
bool is true
in is false
```

As you can see **only inside that function** I'm replacing the input value with a new value. That cannot be done if the input variable is declared as final.

Another reason to do it is so you can pass that input parameter to a inner class/function you can create.
So basically in this specific context it's basically useless, but it changes nothing anyway, so ¯\\\_(ツ)\_/¯
